src/context.rs

```diff
@@ -0,0 +1,1088 @@
+                let ptr: *const TPM2B_NONCE = &TPM2B_NONCE::try_from(val)?;
+                ptr
+            }
+            None => null(),
+        };
+
+        let mut sess = ESYS_TR_NONE;
+
+        let ret = unsafe {
+            Esys_StartAuthSession(
+                self.mut_context(),
+                tpm_key,
+                bind,
+                self.sessions.0,
+                self.sessions.1,
+                self.sessions.2,
+                nonce_ptr,
+                session_type,
+                &symmetric,
+                auth_hash,
+                &mut sess,
+            )
+        };
+
+        let ret = Error::from_tss_rc(ret);
+        if ret.is_success() {
+            let _ = self.open_handles.insert(sess);
+            Ok(sess)
+        } else {
+            error!("Error when creating a session: {}.", ret);
+            Err(ret)
+        }
+    }
+
+    pub fn set_sessions(&mut self, session_handles: (ESYS_TR, ESYS_TR, ESYS_TR)) {
+        self.sessions = session_handles;
+    }
+
+    pub fn sessions(&self) -> (ESYS_TR, ESYS_TR, ESYS_TR) {
+        self.sessions
+    }
+
+    /// Create a primary key and return the handle.
+    ///
+    /// The authentication value, initial data, outside info and creation PCRs are passed as slices
+    /// which are then converted by the method into TSS native structures.
+    ///
+    /// # Constraints
+    /// * `outside_info` must be at most 64 elements long
+    /// * `creation_pcrs` must be at most 16 elements long
+    /// * `auth_value` must be at most 64 elements long
+    /// * `initial_data` must be at most 256 elements long
+    ///
+    /// # Errors
+    /// * if either of the slices is larger than the maximum size of the native objects, a
+    /// `WrongParamSize` wrapper error is returned
+    // TODO: Fix when compacting the arguments into a struct
+    #[allow(clippy::too_many_arguments)]
+    pub fn create_primary_key(
+        &mut self,
+        primary_handle: ESYS_TR,
+        public: &TPM2B_PUBLIC,
+        auth_value: Option<Auth>,
+        initial_data: Option<SensitiveData>,
+        outside_info: Option<Data>,
+        creation_pcrs: &[TPMS_PCR_SELECTION],
+    ) -> Result<ESYS_TR> {
+        let sensitive_create = TPM2B_SENSITIVE_CREATE {
+            size: std::mem::size_of::<TPMS_SENSITIVE_CREATE>()
+                .try_into()
+                .unwrap(), // will not fail on targets of at least 16 bits
+            sensitive: TPMS_SENSITIVE_CREATE {
+                userAuth: match auth_value {
+                    Some(val) => TPM2B_AUTH::try_from(val)?,
+                    None => Default::default(), //wrap_buffer!(auth_value, TPM2B_AUTH, 64),
+                },
+                data: match initial_data {
+                    Some(val) => TPM2B_SENSITIVE_DATA::try_from(val)?,
+                    None => Default::default(), //TPM2B_SENSITIVE_DATA::try_from(initial_data)?,//wrap_buffer!(initial_data, TPM2B_SENSITIVE_DATA, 256),
+                },
```

Comment:
I am not against making a macro. Though I think it should be possible with a one-liner  ... though I failed to do it so I just let it be and made it work "as is". And parahaps improve it later.

Reply:
I was thinking we could do something like this(this should then be applied to all the new types):
```rust
impl Default for SensitiveData {
    fn default() -> Self {
        SensitiveData { value: Vec::<u8>::new() }
    }
}

impl TryFrom<SensitiveData> for TPM2B_SENSITIVE_DATA {
    type Error = Error;
    fn try_from(sensitive_data: SensitiveData) -> Result<Self> {
        if sensitive_data.value.is_empty() {
            return Ok(Default::default());
        }
        if sensitive_data.value.len() > 256 {
            error!("Error: Invalid SensitiveData size(> {})", 256);
            return Err(Error::local_error(WrapperErrorKind::WrongParamSize));
        }
        let mut tss_sensitive_data: TPM2B_SENSITIVE_DATA = Default::default();
        tss_sensitive_data.size = sensitive_data.value().len() as u16;
        tss_sensitive_data.buffer[..sensitive_data.value().len()]
            .copy_from_slice(&sensitive_data.value());
        Ok(tss_sensitive_data)
    }
}
```

Then it would be possible to do:
```rust
                data: TPM2B_SENSITIVE_DATA::try_from(initial_data.unwrap_or_default())?,
```


What do you think about that?